### PR TITLE
Update GetMultiplicity function (E^T * E * 1) to work for H(div) space

### PR DESCRIPTION
**Pull request by rezgarshakeri**

### CODE REVIEW

**Review by jedbrown — APPROVED** [review]
Could use the vector `mult` so you don't need to create another `l_vec`. Also, I think the first apply will overwrite `e_vec` so it doesn't need to be zeroed.

**Comment by rezgarshakeri** [comment]
> 

Right I deleted `l_vec`, but we need to set it to zero (I tried without setting it) before applying restriction otherwise the output values is not correct.

File: interface/ceed-elemrestriction.c [hunk 1]
```diff
@@ -866,15 +866,17 @@ int CeedElemRestrictionGetMultiplicity(CeedElemRestriction rstr,
   int ierr;
   CeedVector e_vec;
 
-  // Create and set e_vec
+  // Create and set l_vec, e_vec
   ierr = CeedElemRestrictionCreateVector(rstr, NULL, &e_vec); CeedChk(ierr);
-  ierr = CeedVectorSetValue(e_vec, 1.0); CeedChk(ierr);
-  ierr = CeedVectorSetValue(mult, 0.0); CeedChk(ierr);
+  ierr = CeedVectorSetValue(mult, 1.0); CeedChk(ierr);
+  ierr = CeedVectorSetValue(e_vec, 0.0); CeedChk(ierr);
```
Comment (jedbrown) [inline interface/ceed-elemrestriction.c:872]: Restrictions overwrite, transpose restrictions add. So this one isn't needed.
```suggestion
```

File: interface/ceed-elemrestriction.c [hunk 2]
```diff
@@ -866,15 +866,17 @@ int CeedElemRestrictionGetMultiplicity(CeedElemRestriction rstr,
...
 
   // Apply to get multiplicity
+  ierr = CeedElemRestrictionApply(rstr, CEED_NOTRANSPOSE, mult, e_vec,
+                                  CEED_REQUEST_IMMEDIATE); CeedChk(ierr);
+  ierr = CeedVectorSetValue(mult, 0.0); CeedChk(ierr);
```
Comment (jedbrown) [inline interface/ceed-elemrestriction.c:877]: And you're correct that this one is needed.

File: interface/ceed-elemrestriction.c [hunk 3]
```diff
@@ -866,15 +866,17 @@ int CeedElemRestrictionGetMultiplicity(CeedElemRestriction rstr,
   int ierr;
   CeedVector e_vec;
 
-  // Create and set e_vec
+  // Create and set l_vec, e_vec
```
Comment (jedbrown) [inline interface/ceed-elemrestriction.c:869]:
```suggestion
  // Create e_vec to hold intermediate computation in E^T (E 1)
```

**Review by jedbrown — APPROVED** [review]

File: interface/ceed-elemrestriction.c [hunk 4]
```diff
@@ -866,15 +866,16 @@ int CeedElemRestrictionGetMultiplicity(CeedElemRestriction rstr,
   int ierr;
   CeedVector e_vec;
 
-  // Create and set e_vec
+  // Create e_vec to hold intermediate computation in E^T (E 1)
   ierr = CeedElemRestrictionCreateVector(rstr, NULL, &e_vec); CeedChk(ierr);
-  ierr = CeedVectorSetValue(e_vec, 1.0); CeedChk(ierr);
-  ierr = CeedVectorSetValue(mult, 0.0); CeedChk(ierr);
+  ierr = CeedVectorSetValue(mult, 1.0); CeedChk(ierr);
```
Comment (jeremylt) [inline interface/ceed-elemrestriction.c:871]: Minor quibble - the comment above now doesn't really apply to both lines here. You could adjust the comment or just move line 871 down to the next block of code?

**Review by jeremylt — APPROVED** [review]

File: interface/ceed-elemrestriction.c (hunk 4 above)
Comment (rezgarshakeri) [reply interface/ceed-elemrestriction.c]: Thanks, How about now?

File: interface/ceed-elemrestriction.c [hunk 5]
```diff
@@ -868,11 +868,12 @@ int CeedElemRestrictionGetMultiplicity(CeedElemRestriction rstr,
...
   ierr = CeedElemRestrictionCreateVector(rstr, NULL, &e_vec); CeedChk(ierr);
-  ierr = CeedVectorSetValue(mult, 1.0); CeedChk(ierr);
 
-  // Apply to get multiplicity
+  // Compute e_vec = E * 1
+  ierr = CeedVectorSetValue(mult, 1.0); CeedChk(ierr);
   ierr = CeedElemRestrictionApply(rstr, CEED_NOTRANSPOSE, mult, e_vec,
                                   CEED_REQUEST_IMMEDIATE); CeedChk(ierr);
+  // Compute multiplicity, mult = E^T * e_vec = E^T (E 1)
   ierr = CeedVectorSetValue(mult, 0.0); CeedChk(ierr);
```
Comment (rezgarshakeri) [inline interface/ceed-elemrestriction.c:877]: @jeremylt

Comment (jeremylt) [reply interface/ceed-elemrestriction.c:877]: LGTM, thanks!

**Pull request merged**